internal/db_impl/sqlite3/query_utils.go

```diff
@@ -0,0 +1,214 @@
+	rows, err := qw.QueryContext(ctx, args...)
+	if err != nil {
+		return nil, mapSQLError(err)
+	}
+
+	return mapSQLRowsFn(rows, m)
+}
+
+func MapStmtRows[T any](ctx context.Context, qw StmtWrapper, args ...any) ([]T, error) {
+	return MapStmtRowsFn(ctx, qw, func(scanner RowScanner) (T, error) {
+		var v T
+
+		err := scanner.Scan(&v)
+
+		return v, err
+	}, args...)
+}
+
+func MapStmtRowFn[T any](ctx context.Context, qw StmtWrapper, m func(RowScanner) (T, error), args ...any) (T, error) {
+	rows := qw.QueryRowContext(ctx, args...)
+
+	return mapSQLRowFn(rows, m)
+}
+
+func MapStmtRow[T any](ctx context.Context, qw StmtWrapper, args ...any) (T, error) {
+	return MapStmtRowFn(ctx, qw, func(scanner RowScanner) (T, error) {
+		var v T
+
+		err := scanner.Scan(&v)
+
+		return v, err
+	}, args...)
+}
+
+func MapQueryRowsFn[T any](ctx context.Context, qw QueryWrapper, query string, m func(RowScanner) (T, error), args ...any) ([]T, error) {
+	rows, err := qw.QueryContext(ctx, query, args...)
+	if err != nil {
+		return nil, mapSQLError(err)
+	}
+
+	return mapSQLRowsFn(rows, m)
+}
+
+func MapQueryRows[T any](ctx context.Context, qw QueryWrapper, query string, args ...any) ([]T, error) {
+	return MapQueryRowsFn(ctx, qw, query, func(scanner RowScanner) (T, error) {
+		var v T
+
+		err := scanner.Scan(&v)
+
+		return v, err
+	}, args...)
+}
+
+func MapQueryRowFn[T any](ctx context.Context, qw QueryWrapper, query string, m func(RowScanner) (T, error), args ...any) (T, error) {
+	row := qw.QueryRowContext(ctx, query, args...)
+
+	return mapSQLRowFn(row, m)
+}
+
+func MapQueryRow[T any](ctx context.Context, qw QueryWrapper, query string, args ...any) (T, error) {
+	return MapQueryRowFn(ctx, qw, query, func(scanner RowScanner) (T, error) {
+		var v T
+
+		err := scanner.Scan(&v)
+
+		return v, err
+	}, args...)
+}
+
+func ExecQueryAndCheckUpdatedNotZero(ctx context.Context, wrapper QueryWrapper, query string, args ...any) error {
+	updated, err := ExecQuery(ctx, wrapper, query, args...)
+	if err != nil {
+		return err
+	}
+
+	if updated == 0 {
+		return fmt.Errorf("no values changed")
+	}
+
+	return nil
+}
+
+func ExecStmtAndCheckUpdatedNotZero(ctx context.Context, wrapper StmtWrapper, args ...any) error {
+	updated, err := ExecStmt(ctx, wrapper, args...)
+	if err != nil {
+		return err
+	}
+
+	if updated == 0 {
+		return fmt.Errorf("no values changed")
+	}
+
+	return nil
+}
+
+func ExecQuery(ctx context.Context, wrapper QueryWrapper, query string, args ...any) (int, error) {
+	r, err := wrapper.ExecContext(ctx, query, args...)
+	if err != nil {
+		return 0, err
+	}
+
+	affected, err := r.RowsAffected()
+	if err != nil {
+		panic("affected rows is unsupported")
+	}
+
+	return int(affected), nil
+}
+
+func ExecStmt(ctx context.Context, wrapper StmtWrapper, args ...any) (int, error) {
+	r, err := wrapper.ExecContext(ctx, args...)
+	if err != nil {
+		return 0, err
+	}
+
+	affected, err := r.RowsAffected()
+	if err != nil {
+		panic("affected rows is unsupported")
```

Comment:
wouldn't be useful to know the error message here?

Reply:
This is a particularity of the underlying DB implementation. SQLite supports this. We *should* never see this panic unless something went really wrong.